### PR TITLE
Generic registry support

### PR DESCRIPTION
**Pull request by Relrin**
The following pull request covers the initial steps for supporting communication between the existing actors (issue #158 ) and hot-reload features (issue #103). 

I have plans to add tests and a documentation for this PR. However I would like to request a small review for this code first, because perhaps some changes will be needed. And after it keep going with other aspects of implementation.

##### Checklist

- [x] tests are passing with `cargo test`. 
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message is clear
- [x] the example of the registry usage

### CODE REVIEW

**Comment by Relrin** [comment]
I also thinking about the key/value, that stored in `dashmap`: is it has a sense to replace it on something else (e.g. `ChildRef` for key)?

**Comment by o0Ignition0o** [comment]
Hey there !
It looks great ! I'll try to have a look at it tonight or tomorrow :)

**Comment by Relrin** [comment]
I added the `broadcast_message` method for the `RegistryDispatcher` trait so that it will be possible to pass desired messages to registered actors as a developer would like. Also appended the docs strings for the structs and enums there.

The main idea for this public API is to expose a couple of method for the global place with registries so it could pass messages, register and remove actors to the desired registry as is. 
Basically it means that each actor which wants to broadcast a message, will use the "global call" with passing a registry name / type and a message for building a desired pipeline for communication globally or with the desired group of actors.

**Comment by o0Ignition0o** [comment]
Hey there,

I'm having a bit of a hard time to reason with the naming and how the Registry / Dispatcher behaves.

The main thing I'm trying to make sure is that we scope it well (like we dont expose to much internals).

I'll have a much more thorough look at it tonight, starting with the comments :)

**Comment by Relrin** [comment]
> I'm having a bit of a hard time to reason with the naming and how the Registry / Dispatcher behaves.
> 
> The main thing I'm trying to make sure is that we scope it well (like we dont expose to much internals).

Okay, let me share some ideas how I see these changes. Let's start from handling registries in general (on the system level):
1) The [GlobalSystem](https://github.com/bastion-rs/bastion/blob/8f6dc5a3387e742b7815945efad3c38aebefb864/bastion/src/system.rs#L23-L31) struct will have an additional field that will store all used registries. To make it fast, it could be declared as the additional hashmap, for example as the `DHashMap<RegistryType, Registry>` type. The key is the name of registry and the values in this pair is the real DHashMap (perhaps as the reference?) that represents actors which are organized in the certain group with any desired dispatcher implementation.
2) The registry must be declared as the part of a child definition with the help of the special `.with_registry` method. For example:  
    ```rust
   fn actors_with_registry(children: Children) -> Children {
        // ...

        children
            .with_registry(
                &Registry::default()
                    .with_registry_type(RegistryType::Named("my_group"))
                    .with_dispatcher(Box::new(SelectRandomlyDispatcher::new()))
            )
            // Declare the second registry that you want to use here
            .with_registry( 
                // ... 
            )
            .with_exec(move |_ctx: BastionContext| async move {
                // do something...
            })
   } 
    ```
This behaviour gives two main opportunities: 
- Declare the used registry just in place with the used actors
- Will add the new registry to the `GlobalSystem` internally (via the private API) if it hasn't added yet. If registries in separate actor declarations have the same the same name, basically it means that they should be organized under the same registry.

P.S. Passing registries by a reference gives an opportunity to organize different children in the same registry and do notification as you would like. 

The next part is what available for the actors itself:
1. They can declare on which registries they want to join (as was mentioned before, use the `.with_registry` method).
2. Before starting the actor we will call private API for adding the actor to the declared registries just via the simple for loop. It will be implemented something like this (as the part of the actor's codebase):
```rust 
pub(crate) fn register_actor {
   for registry in self.registries {
       SYSTEM.register_node(registry.registry_type, ctx.current(), module_path!())
   }
} 

```
3. The similar stuff applies for the case when we need to unregister the failed/stopped/killed actor: just pass through the given registries an remove the entries from the declared registries.
4. Any actor will have a new method for communication via Registries. Let's say it will be named as the `registry_broadcast`, and will require specifying the `RegistryType` value with the message of the `SignedMessage` type that you want to pass to dispatcher that will handle the rest of staff. Internally, this `registry_broadcast` method will pass the message to the `GlobalSystem` instance so that it could deliver the message to the certain group in registry (or just drop the message if it can't be delivered due unavailability).

Any ideas to make it better are welcome :)

**Comment by o0Ignition0o** [comment]
Thanks for such a thorough explanation.

Let’s make sure I got the intention right:

Put a registry index in the GlobalSystem so we can keep the names and handles to each available registry. The registries will be indexed by name.

Allow an actor to be « bound » to a given registry. (So they can listen to registry broadcasts)

Provide a broadcast method to each actor.

Provide a dispatch method to a registry to define the way a registry routes messages to actors.

^ This is the part I’m not really sure I understood correctly. What is dispatch supposed to do ?
 
Thank’s for your time :)

**Comment by Relrin** [comment]
> Put a registry index in the GlobalSystem so we can keep the names and handles to each available registry. The registries will be indexed by name.

Yep, that's correct. 

> Allow an actor to be « bound » to a given registry. (So they can listen to registry broadcasts)

Correct. And more over, they could react somehow on the "event" when someone added to registry. But perhaps then it make a sense to add another method but for removing an actor from registry. 🤔 

> Provide a broadcast method to each actor.

Basically it means the legal / public way to communicate with the whole group of actors that were registered in the certain registry. It super useful thing, for example for the cases when the actor who doesn't belong to the certain group could send a message to the desired group with the useful information or want to pass the extracted piece of data to the next stage in your processing pipeline.

However, the logic of how and to whom to send a passed message must be defined in implementation with the help of `RegistryDispatcher` trait.

> ^ This is the part I’m not really sure I understood correctly. What is dispatch supposed to do ?

Maybe event better to rename it to the `notify` instead of `dispatch` to clarify the purpose of this method. This method is the part of the `RegistryDispatcher` trait that invokes right after adding the actor to group in the certain registry.
This logic can be used for the cases when you want to do specific logic right after getting knowledge about that someone else was added to the group.

**Comment by o0Ignition0o** [comment]
> Basically it means the legal / public way to communicate with the whole group of actors that were registered in the certain registry.

Oh that's great indeed. Are there conventions that make it behave so? I don't have a lot of domain knowledge, but I would have maybe considered an enum like (pseudocode):

```rust
child.notify(Recipient::Actor("an_other_actor"), message);
child.notify(Recipient::Registry("a_registry_i_dont_belong_to"), message);
```

> This method is the part of the RegistryDispatcher trait that invokes right after adding the actor to group in the certain registry.

So the "javascript" way to name `dispatch` it would be `onNewActor()` ?

I never know how to name these correctly, sometimes i go with `new_actor_handler` or `handle_new_actor`, or `on_new_actor`... I m not really sure ^^' but the (really great!) explanation you gave makes me think we wanna convey the meaning in the method name.

If that's the case that would probably allow us to clear the following concern:

> they could react somehow on the "event" when someone added to registry.

In pseudo code that could look like:

```rust
/// Method that handles logic with registering a new actor.
fn my_super_dispatch(&self, entries: &BastionRegistryMap) {
	for (id, name) in entries.iter() {
		self.notify(Recipient::Actor(name), "new player!");
	}
}
```

I'm still not sure I understood everything correctly, please let me know if not || if it doesn't make sense ^^

**Comment by vertexclique** [comment]
Ok, so after a rough look at the comment and the registry implementation. `Registry` as a name shouldn't be seen to the user. That is making life quite a bit complicated. So that said:
* I would like to keep the `registry` as an internal name, which is holding the dispatching routes.
* I would prefer to have `Dispatcher` configured with the public API like how you pointed out.
     * If that's not configured, use the `DefaultDispatcher` that is using the `Default` registry.

The ideas above are coming from how Akka works with their Dispatchers. To understand more please head to:
https://doc.akka.io/docs/akka/current/typed/dispatchers.html

**Comment by Relrin** [comment]
> Oh that's great indeed. Are there conventions that make it behave so? I don't have a lot of domain knowledge, but I would have maybe considered an enum like (pseudocode):
> 
> ```rust
> self.notify(Recipient::Actor("an_other_actor"), message);
> self.notify(Recipient::Registry("a_registry_i_dont_belong_to"), message);
> ```
> 

There's no a universal way to do it, so I'm just trying to implement a new functionality based on the existing codebase without breaking anything. :)

The main idea for the `notify` call is somehow to say other actors in the certain registry that someone joined to or removed from the registry. How to react on this type of events, for example send some piece of data for each other actor in the same registry or another one, relies completely on developers implementation.

In the same time the purpose of the `broadcast_message` method is to pass the message to the given registry with an appropriate message (which is must be of the `SignedMessage` type that let us to understand from whom the message).

> So the "javascript" way to name `dispatch` would be `onNewActor()` ?
> 
> I never know how to name these correctly, sometimes i go with `new_actor_handler` or `handle_new_actor`, or `on_new_actor`... I m not really sure ^^' but the (really great!) explanation you gave makes me think we wanna convey the meaning in the method name.

How to name handlers for each particular type of events defined by the conventions in your own project. But in the most projects that I was working or faced in the past, quite often started with `On*` in names for events to keep the codebase clean and readable for others. 

I'm not sure that we need the similar convention, because bastion has the `handle` method, which is used for handling those cases (or similar to) :)

Regarding the last piece of code:
```rust
/// Method that lets everyone know there's a new wolf in the pack!
fn dispatch_new_player(&self, entries: &BastionRegistryMap) {
	for (id, name) in entries.iter() {
		self.notify(Recipient::Actor(name), "new player!");
	}
}
```
At the moment there's no assigned names for actors / supervisors / registries. But it probably has a sense to add a field for this case. The single corner case about I'm thinking about is to how to sent directly a message to an actor (or anything else) by the name without the usage of registry (like how it's done in Elixir / Erlang)

**Comment by Relrin** [comment]
> Ok, so after a rough look at the comment and the registry implementation. `Registry` as a name shouldn't be seen to the user. That is making life quite a bit complicated. So that said:
> 
> * I would like to keep the `registry` as an internal name, which is holding the dispatching routes.
> * I would prefer to have `Dispatcher` configured with the public API like how you pointed out.
>   
>   * If that's not configured, use the `DefaultDispatcher` that is using the `Default` registry.
> 
> The ideas above are coming from how Akka works with their Dispatchers. To understand more please head to:
> https://doc.akka.io/docs/akka/current/typed/dispatchers.html

It could be renamed to the `Dispatcher` instead of  the `Registry` usage in the project. As a developer I only want to configure the dispatcher, and attach to the cluster as is. In the same time, I don't want to impact on this dispatcher directly, but want to provide minimalistic API for the later usage (like the notify someone from the group in registry, and react on join/remove)

**Comment by vertexclique** [comment]
> It could be renamed to the `Dispatcher` instead of the `Registry` usage in the project.
Yes and thank you.

> As a developer I only want to configure the dispatcher, and attach to the cluster as is. In the same time, I don't want to impact on this dispatcher directly, but want to provide minimalistic API for the later usage (like the send the message to someone from the group in registry, and react on join/remove)

Let me give one of my use cases also: I want a pool of actors that will have a different kind of dispatching mechanism to their mailboxes. That said I want to implement:
* Send incoming message to the actor in the pool which has a smaller mailbox size. **OR**
* Work steal from one of the other actors.

Moreover, I am responding to the high workload and want to increase the throughput by autoscale the actors. Then I want a dynamically resizing pool backed by an actor spawning dispatcher on hog.

**Comment by Relrin** [comment]
> Let me give one of my use cases also: I want a pool of actors that will have a different kind of dispatching mechanism to their mailboxes. That said I want to implement:
> 
> * Send incoming message to the actor in the pool which has a smaller mailbox size. **OR**
> * Work steal from one of the other actors.
> 
> Moreover, I am responding to the high workload and want to increase the throughput by autoscale the actors. Then I want a dynamically resizing pool backed by an actor spawning dispatcher on hog.

It's a quite interesting idea. However, it means that the dispatcher must be attached to / must be a part of the ChildrenRef struct (that holds the all entire group of the actors), so it could distribute tasks/message based on some strategy, isn't it?

This case can be described with the following example:
1) We have actors `A` and `B` which are belong different groups (or in the same, it doesn't matter for this example).
2) Actor `A` sends the message to global registry with the signature something like:
```rust
self.broadcast_message(RegistryType::Named("Group X"), message);
```
3) Global registry retrieving passing the message to the certain registry with a given name.
4) Registry with the `Group X` name passed the message to the supervisor with a `Dispatcher`.
5) Dispatcher based on the certain logic passed the message to the Actor `B`  

In graph it will look like:
Actor A  -> Global Registry -> Registry with the `Group X` name  -> ChildrenRef -> Dispatcher  -> Actor B

However, what ever the case we have, the entry point for dispatching require to have a global registry that holds references of all available actors.

Correct me, please, if I made a mistake or misunderstood.

**Comment by vertexclique** [comment]
> It's a quite interesting idea. However, it means that the dispatcher must be attached to / must be a part of the ChildrenRef struct (that holds the all entire group of the actors), so it could distribute tasks/message based on some strategy, isn't it?

Either that or Dispatcher encapsulates all `*Ref`'s and wraps their API.


> Actor A sends the message to global registry with the signature something like:
```
self.broadcast_message(RegistryType::Named("Group X"), message);
```

Forget about showing `RegistryType` to users call it dispatchers. The method probably be like:
```
self.broadcast_message(Dispatcher::Name("Group X"), message);
```

Then this will happen:
Actor A -> Global Dispatcher -> Global Registry -> Registry with the name `Group X` -> Get the Group X Dispatcher -> ChildrenRef -> Actor B

At least this is what I have in my mind. We can refactor this if there are pitfalls.

File: bastion/src/dispatcher.rs [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+    /// Creates a new instance of the global registry.
+    pub(crate) fn new() -> Self {
+        GlobalDispatcher {
+            dispatchers: DashMap::new(),
+        }
+    }
+
+    /// Passes the notification from the actor to everyone that registered in the same
+    /// groups as the caller.
+    pub(crate) fn notify(
```
Comment (vertexclique) [inline bastion/src/dispatcher.rs]: #[allow(dead_code)]

File: bastion/src/dispatcher.rs [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+//! actors grouped together.
+use crate::child_ref::ChildRef;
+use crate::envelope::SignedMessage;
+use dashmap::DashMap;
+use std::fmt::{self, Debug};
+use std::hash::{Hash, Hasher};
+
+/// Type alias for the concurrency hashmap. Each key-value pair stores
+/// the Bastion identifier as the key and the module name as the value.
+pub type BastionDispatcherMap = DashMap<ChildRef, String>;
```
Comment (vertexclique) [inline bastion/src/dispatcher.rs]: Let's call this `DispatcherMap`.

File: bastion/src/dispatcher.rs (hunk 1 above)
Comment (Relrin) [reply bastion/src/dispatcher.rs]: Fixed

File: bastion/src/dispatcher.rs (hunk 2 above)
Comment (Relrin) [reply bastion/src/dispatcher.rs]: Done

File: bastion/src/system.rs [hunk 3]
```diff
@@ -154,10 +162,12 @@ impl System {
...
         root_sv.children_with_id(NIL_ID, |children| {
-            children.with_exec(|ctx: BastionContext| async move {
-                loop {
-                    let smsg = ctx.recv().await?;
-                    debug!("Received dead letter: {:?}", smsg);
+            children.with_exec(|ctx: BastionContext| {
+                async move {
+                    loop {
+                        let smsg = ctx.recv().await?;
+                        debug!("Received dead letter: {:?}", smsg);
```
Comment (vertexclique) [inline bastion/src/system.rs]: If this is here, the next task for us putting a callback here for the dead letters :) 🚀 Noice!

File: bastion/examples/broadcast_message.rs [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+///    to the next stage with the similar call to the Reduce group.
+/// 3. The actor from the Response group retrieves the data from the actors of the
+///    Reduce group, combines the results and prints its when everything is done.
+///
+fn main() {
+    Bastion::init();
+
+    Bastion::supervisor(input_supervisor).expect("Couldn't create the supervisor.");
+    Bastion::supervisor(map_supervisor).expect("Couldn't create the supervisor.");
+    Bastion::supervisor(response_supervisor).expect("Couldn't create the supervisor.");
```
[the remark this answers is not in the document]
Comment (Relrin) [reply bastion/examples/broadcast_message.rs]: Fixed
